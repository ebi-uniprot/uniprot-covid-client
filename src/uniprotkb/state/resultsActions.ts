import { action } from 'typesafe-actions';
import { Dispatch, Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import fetchData from '../../shared/utils/fetchData';
import { RootState } from '../../app/state/rootInitialState';
import apiUrls from '../config/apiUrls';
import { Column } from '../types/columnTypes';
import {
  ColumnSelectTab,
  FieldDatum,
  ReceivedFieldData,
  ReceivedField,
} from '../types/resultsTypes';

export const UPDATE_COLUMN_SORT = 'UPDATE_COLUMN_SORT';
export const SWITCH_VIEW_MODE = 'SWITCH_VIEW_MODE';
export const RECEIVE_FIELDS = 'RECEIVE_FIELDS';
export const REQUEST_FIELDS = 'REQUEST_FIELDS';
export const UPDATE_TABLE_COLUMNS = 'UPDATE_TABLE_COLUMNS';

export const requestFields = () => action(REQUEST_FIELDS);

export const prepareFields = (fields: ReceivedField[]) =>
  fields.map(({ label, name }) => ({ id: name as Column, label }));

export const prepareFieldData = (fieldData: ReceivedFieldData) => {
  const dataTab: FieldDatum[] = [];
  const linksTab: FieldDatum[] = [];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const linksAdded: any = {};
  fieldData.forEach(({ groupName, fields, isDatabaseGroup, id }) => {
    const group = {
      id,
      title: groupName,
      items: prepareFields(fields),
    };
    if (isDatabaseGroup) {
      if (!linksAdded[groupName]) {
        linksTab.push(group);
        linksAdded[groupName] = true;
      }
    } else {
      dataTab.push(group);
    }
  });
  return {
    [ColumnSelectTab.data]: dataTab,
    [ColumnSelectTab.links]: linksTab,
  };
};
export const receiveFields = (data: ReceivedFieldData) =>
  action(RECEIVE_FIELDS, {
    data: prepareFieldData(data),
    receivedAt: Date.now(),
  });

export const fetchFields = () => async (dispatch: Dispatch) => {
  dispatch(requestFields());
  return fetchData<ReceivedFieldData>(apiUrls.resultsFields).then((response) =>
    dispatch(receiveFields(response.data))
  );
};

export const shouldFetchFields = (state: RootState) => {
  const { fields } = state.results;
  return (
    !fields.isFetching &&
    (!fields.data.data.length || !fields.data.links.length)
  );
};

export const fetchFieldsIfNeeded = () => (
  dispatch: ThunkDispatch<RootState, void, Action>,
  getState: () => RootState
) => {
  if (shouldFetchFields(getState())) {
    dispatch(fetchFields());
  }
};

export const updateTableColumns = (tableColumns: Column[]) =>
  action(UPDATE_TABLE_COLUMNS, {
    tableColumns,
  });
