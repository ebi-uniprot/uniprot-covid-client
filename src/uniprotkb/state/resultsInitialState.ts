import { Column } from '../types/columnTypes';
import { FieldData, ColumnSelectTab } from '../types/resultsTypes';

export enum ViewMode {
  TABLE,
  CARD,
}

export type ResultsState = {
  tableColumns: Column[];
  fields: {
    data: FieldData;
    isFetching: boolean;
  };
};

export const defaultTableColumns = [
  Column.accession,
  Column.reviewed,
  Column.id,
  Column.proteinName,
  Column.geneNames,
  Column.organismName,
];

const resultsInitialState = {
  tableColumns: defaultTableColumns,
  fields: {
    data: {
      [ColumnSelectTab.data]: [],
      [ColumnSelectTab.links]: [],
    },
    isFetching: false,
  },
};

export default resultsInitialState;
