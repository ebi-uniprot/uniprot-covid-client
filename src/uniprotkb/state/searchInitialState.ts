import { createPreSelectedClauses } from '../utils/clause';
import {
  Clause,
  Namespace,
  Evidence,
  Evidences,
  SearchTermType,
} from '../types/searchTypes';

export type SearchState = {
  readonly clauses: Clause[];
  readonly namespace: Namespace;
  readonly searchTerms: { data: SearchTermType[]; isFetching: boolean };
  readonly evidences: Evidences;
};

const searchInitialState: SearchState = {
  clauses: createPreSelectedClauses(),
  namespace: Namespace.uniprotkb,
  searchTerms: {
    data: [],
    isFetching: false,
  },
  evidences: {
    [Evidence.GO]: {
      data: [],
      isFetching: false,
    },
    [Evidence.ANNOTATION]: {
      data: [],
      isFetching: false,
    },
  },
};

export default searchInitialState;
