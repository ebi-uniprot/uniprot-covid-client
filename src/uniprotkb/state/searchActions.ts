import { Dispatch, Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from 'typesafe-actions';
import fetchData from '../../shared/utils/fetchData';
import apiUrls from '../config/apiUrls';
import {
  SearchTermType,
  Operator,
  Evidence,
  Clause,
  dataType,
} from '../types/searchTypes';
import { RootState } from '../../app/state/rootInitialState';

export const SELECT_SEARCH_TERM = 'SELECT_SEARCH_TERM';
export const UPDATE_INPUT_VALUE = 'UPDATE_INPUT_VALUE';
export const UPDATE_EVIDENCE = 'UPDATE_EVIDENCE';
export const UPDATE_RANGE_VALUE = 'UPDATE_RANGE_VALUE';
export const UPDATE_LOGIC_OPERATOR = 'UPDATE_LOGIC_OPERATOR';
export const HANDLE_FIELD_SELECT = 'HANDLE_FIELD_SELECT';
export const SUBMIT_ADVANCED_QUERY = 'SUBMIT_ADVANCED_QUERY';
export const SET_PRE_SELECTED_CLAUSES = 'SET_PRE_SELECTED_CLAUSES';
export const ADD_CLAUSE = 'ADD_CLAUSE';
export const REMOVE_CLAUSE = 'REMOVE_CLAUSE';
export const REQUEST_SEARCH_TERMS = 'REQUEST_SEARCH_TERMS';
export const RECEIVE_SEARCH_TERMS = 'RECEIVE_SEARCH_TERMS';
export const REQUEST_EVIDENCES = 'REQUEST_EVIDENCES';
export const RECEIVE_EVIDENCES = 'RECEIVE_EVIDENCES';
export const UPDATE_CLAUSES = 'UPDATE_CLAUSES';
export const UPDATE_QUERY_STRING = 'UPDATE_QUERY_STRING';
export const RESET = 'RESET';
export const RESET_SEARCH_INPUT = 'RESET_SEARCH_INPUT';

export const selectSearchTerm = (
  clauseId: string,
  searchTerm: SearchTermType
) =>
  action(SELECT_SEARCH_TERM, {
    clauseId,
    searchTerm,
    queryInput:
      searchTerm.dataType === dataType.enum &&
      searchTerm.values &&
      searchTerm.values.length
        ? { stringValue: searchTerm.values[0].value }
        : {},
  });

export const updateInputValue = (
  clauseId: string,
  value: string,
  id?: string
) =>
  action(UPDATE_INPUT_VALUE, {
    clauseId,
    value,
    id,
  });

export const updateEvidence = (clauseId: string, value: string) =>
  action(UPDATE_EVIDENCE, {
    clauseId,
    value,
  });

export const updateRangeValue = (
  clauseId: string,
  value: string,
  from?: boolean
) =>
  action(UPDATE_RANGE_VALUE, {
    clauseId,
    value,
    from,
  });

export const updateLogicOperator = (clauseId: string, value: Operator) =>
  action(UPDATE_LOGIC_OPERATOR, {
    clauseId,
    value,
  });

export const submitAdvancedQuery = () => action(SUBMIT_ADVANCED_QUERY);

export const setPreSelectedClauses = () => action(SET_PRE_SELECTED_CLAUSES);

export const addClause = () => action(ADD_CLAUSE);

export const removeClause = (clauseId: string) =>
  action(REMOVE_CLAUSE, {
    clauseId,
  });

export const updateClauses = (clauses: Clause[]) =>
  action(UPDATE_CLAUSES, {
    clauses,
  });

export const requestSearchTerms = () => action(REQUEST_SEARCH_TERMS);

export const receiveSearchTerms = (data: SearchTermType[]) =>
  action(RECEIVE_SEARCH_TERMS, {
    data,
    receivedAt: Date.now(),
  });

export const fetchSearchTerms = () => async (dispatch: Dispatch) => {
  dispatch(requestSearchTerms());
  return fetchData<SearchTermType[]>(
    apiUrls.advancedSearchTerms
  ).then((response) => dispatch(receiveSearchTerms(response.data)));
};

export const shouldFetchSearchTerms = (state: RootState) => {
  const { searchTerms } = state.query;
  return !searchTerms.isFetching && !searchTerms.data.length;
};

export const fetchSearchTermsIfNeeded = () => (
  dispatch: ThunkDispatch<RootState, void, Action>,
  getState: () => RootState
) => {
  if (shouldFetchSearchTerms(getState())) {
    dispatch(fetchSearchTerms());
  }
};

export const requestEvidences = (evidencesType: Evidence) =>
  action(REQUEST_EVIDENCES, {
    evidencesType,
  });

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const receiveEvidences = (data: any, evidencesType: Evidence) =>
  action(RECEIVE_EVIDENCES, {
    data,
    evidencesType,
    receivedAt: Date.now(),
  });

export const fetchEvidences = (evidencesType: Evidence) => async (
  dispatch: ThunkDispatch<RootState, void, Action>
) => {
  const url = apiUrls.evidences[evidencesType];
  dispatch(requestEvidences(evidencesType));
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return fetchData<any>(url).then((response) =>
    dispatch(receiveEvidences(response.data, evidencesType))
  );
};

export const shouldFetchEvidences = (state: RootState, evidence: Evidence) => {
  const evidences = state.query.evidences[evidence];
  return !evidences.isFetching && !evidences.data.length;
};

export const fetchEvidencesIfNeeded = (evidencesType: Evidence) => (
  dispatch: ThunkDispatch<RootState, void, Action>,
  getState: () => RootState
) => {
  if (shouldFetchEvidences(getState(), evidencesType)) {
    dispatch(fetchEvidences(evidencesType));
  }
};

export const updateQueryString = (queryString: string) =>
  action(UPDATE_QUERY_STRING, {
    queryString,
  });

export const reset = () => action(RESET);

export const resetSearchInput = () => action(RESET_SEARCH_INPUT);
