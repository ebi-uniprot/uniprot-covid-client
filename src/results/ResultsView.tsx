import React from 'react';
import { DataTable, DataList } from 'franklin-sites';
import ColumnConfiguration from '../model/ColumnConfiguration';
import '../styles/alert.scss';
import '../styles/ResultsView.scss';
import { SelectedEntries, SortDirection } from './types/resultsTypes';
import UniProtCard from '../view/uniprotkb/components/UniProtCard';
import uniProtKbConverter, {
  UniProtkbUIModel,
  UniProtkbAPIModel,
} from '../model/uniprotkb/UniProtkbConverter';
import { ViewMode } from './state/resultsInitialState';
import { SortableColumn, Column } from '../model/types/ColumnTypes';

type ResultsTableProps = {
  results: UniProtkbAPIModel[];
  tableColumns: (Column | SortableColumn)[];
  selectedEntries: SelectedEntries;
  handleEntrySelection: (rowId: string) => void;
  handleHeaderClick: (column: SortableColumn) => void;
  handleLoadMoreRows: () => void;
  totalNumberResults: number;
  sortColumn: SortableColumn;
  sortDirection: SortDirection;
  viewMode: ViewMode;
};

const ResultsView: React.FC<ResultsTableProps> = ({
  results = [],
  totalNumberResults,
  tableColumns,
  selectedEntries,
  handleEntrySelection,
  handleLoadMoreRows,
  handleHeaderClick,
  sortColumn,
  sortDirection,
  viewMode,
}) => {
  const hasMoreData = totalNumberResults > results.length;
  if (viewMode === ViewMode.CARD) {
    return (
      <div className="datalist">
        <DataList
          idKey="primaryAccession"
          data={results}
          dataRenderer={(dataItem: UniProtkbAPIModel) => (
            <UniProtCard
              data={dataItem}
              selectedEntries={selectedEntries}
              handleEntrySelection={handleEntrySelection}
            />
          )}
          onLoadMoreItems={handleLoadMoreRows}
          hasMoreData={hasMoreData}
        />
      </div>
    );
  } // viewMode === ViewMode.TABLE
  const columns = tableColumns.map(columnName => {
    const columnConfig = ColumnConfiguration.get(columnName);
    if (columnConfig) {
      return {
        label: columnConfig.label,
        name: columnName,
        render: (row: UniProtkbAPIModel) =>
          columnConfig.render(uniProtKbConverter(row) as UniProtkbUIModel),
        sortable: columnConfig.sortable,
        sorted: columnName === sortColumn && sortDirection,
      };
    }
    return {
      label: columnName,
      name: columnName,
      render: () => (
        <div className="warning">{`${columnName} has no config`}</div>
      ),
      sortable: false,
      sorted: false,
    };
  });
  return (
    <DataTable
      idKey="primaryAccession"
      columns={columns}
      data={results}
      selectable
      selected={selectedEntries}
      onSelect={handleEntrySelection}
      onHeaderClick={handleHeaderClick}
      onLoadMoreItems={handleLoadMoreRows}
      hasMoreData={hasMoreData}
    />
  );
};

export default ResultsView;
