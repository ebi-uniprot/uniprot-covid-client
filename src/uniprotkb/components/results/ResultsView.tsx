import React, { useState, useEffect } from 'react';
import { DataTable, DataList, Loader } from 'franklin-sites';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import ColumnConfiguration from '../../config/ColumnConfiguration';
import { SortDirection } from '../../types/resultsTypes';
import UniProtKBCard from './UniProtKBCard';
import uniProtKbConverter, {
  UniProtkbUIModel,
  UniProtkbAPIModel,
} from '../../adapters/uniProtkbConverter';
import { ViewMode } from '../../state/resultsInitialState';
import { SortableColumn, Column } from '../../types/columnTypes';
import './styles/warning.scss';
import './styles/results-view.scss';
import { getParamsFromURL, getLocationObjForParams } from '../../utils/results-utils';
import { getAPIQueryUrl } from '../../config/apiUrls';
import useDataApi from '../../../shared/hooks/useDataApi';
import getNextUrlFromResponse from '../../utils/queryUtils';
import BaseLayout from '../../../shared/components/layouts/BaseLayout';
import NoResultsPage from '../../../shared/components/error-pages/NoResultsPage';

type ResultsTableProps = {
  selectedEntries: string[];
  columns: Column[];
  viewMode: ViewMode;
  handleEntrySelection: (rowId: string) => void;
} & RouteComponentProps;

const ResultsView: React.FC<ResultsTableProps> = ({
  selectedEntries,
  columns,
  viewMode,
  handleEntrySelection,
  history,
  location,
}) => {
  const { search: queryParamFromUrl } = location;
  const { query, selectedFacets, sortColumn, sortDirection } = getParamsFromURL(
    queryParamFromUrl
  );

  const initialApiUrl = getAPIQueryUrl(
    query,
    columns,
    selectedFacets,
    sortColumn,
    sortDirection
  );

  const [url, setUrl] = useState(initialApiUrl);
  const [metaData, setMetaData] = useState<{
    total: number;
    nextUrl: string | undefined;
  }>({ total: 0, nextUrl: undefined });
  const [allResults, setAllResults] = useState<UniProtkbAPIModel[]>([]);

  const { data, headers } = useDataApi(url);
  // TODO handle error

  useEffect(() => {
    if (!data) {
      return;
    }
    const { results } = data;
    setAllResults((allRes) => [...allRes, ...results]);
    setMetaData(() => ({
      total: headers['x-totalrecords'],
      nextUrl: getNextUrlFromResponse(headers.link),
    }));
  }, [data, headers]);

  if (allResults.length === 0) {
    return <Loader />;
  }

  if (allResults.length === 0) {
    return (
      <BaseLayout>
        <NoResultsPage />
      </BaseLayout>
    );
  }

  const { total, nextUrl } = metaData;

  const handleLoadMoreRows = () => nextUrl && setUrl(nextUrl);

  const updateColumnSort = (column: SortableColumn) => {
    /**
     * NOTE: temporary fix until backend provide
     * the correct name for sort fields
     * https://www.ebi.ac.uk/panda/jira/browse/TRM-23753
     */
    const fieldNameMap = new Map([
      [Column.accession, 'accession'],
      [Column.id, 'mnemonic'],
      [Column.proteinName, 'name'],
      [Column.geneNames, 'gene'],
      [Column.organism, 'organism_name'],
      [Column.mass, 'mass'],
      [Column.length, 'length'],
    ]);
    const apiColumn = fieldNameMap.get(column);

    // Change sort direction
    const updatedSortDirection =
      !sortDirection || sortDirection === SortDirection.descend
        ? SortDirection.ascend
        : SortDirection.descend;

    history.push(
      getLocationObjForParams(
        '/uniprotkb',
        query,
        selectedFacets,
        apiColumn,
        updatedSortDirection
      )
    );
  };

  const hasMoreData = total > allResults.length;
  if (viewMode === ViewMode.CARD) {
    return (
      <div className="datalist">
        <DataList
          getIdKey={({ primaryAccession }: { primaryAccession: string }) =>
            primaryAccession
          }
          data={allResults}
          dataRenderer={(dataItem: UniProtkbAPIModel) => (
            <UniProtKBCard
              data={dataItem}
              selected={selectedEntries.includes(dataItem.primaryAccession)}
              handleEntrySelection={handleEntrySelection}
            />
          )}
          onLoadMoreItems={handleLoadMoreRows}
          hasMoreData={hasMoreData}
          loaderComponent={<Loader />}
        />
      </div>
    );
  } // viewMode === ViewMode.TABLE
  const columnsToDisplay = columns.map((columnName) => {
    const columnConfig = ColumnConfiguration.get(columnName);
    if (columnConfig) {
      return {
        label: columnConfig.label,
        name: columnName,
        render: (row: UniProtkbAPIModel) =>
          columnConfig.render(uniProtKbConverter(row) as UniProtkbUIModel),
        sortable: columnConfig.sortable,
        sorted: columnName === sortColumn && sortDirection, // TODO this doesn't seem to update the view
      };
    }
    return {
      label: columnName,
      name: columnName,
      render: () => (
        <div className="warning">{`${columnName} has no config`}</div>
      ),
      sortable: false,
      sorted: false,
    };
  });
  return (
    <DataTable
      getIdKey={({ primaryAccession }: { primaryAccession: string }) =>
        primaryAccession
      }
      columns={columnsToDisplay}
      data={allResults}
      selectable
      selected={selectedEntries}
      onSelect={handleEntrySelection}
      onHeaderClick={updateColumnSort}
      onLoadMoreItems={handleLoadMoreRows}
      hasMoreData={hasMoreData}
      loaderComponent={<Loader />}
    />
  );
};

export default withRouter(ResultsView);
