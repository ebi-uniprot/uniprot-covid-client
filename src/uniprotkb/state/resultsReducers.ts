import { ActionType } from 'typesafe-actions';
import * as resultsActions from './resultsActions';
import resultsInitialState, { ResultsState } from './resultsInitialState';

export type ResultAction = ActionType<typeof resultsActions>;
const resultsReducers = (
  state: ResultsState = resultsInitialState,
  action: ResultAction
) => {
  switch (action.type) {
    case resultsActions.REQUEST_FIELDS:
      return {
        ...state,
        fields: { ...state.fields, isFetching: true },
      };
    case resultsActions.RECEIVE_FIELDS:
      return {
        ...state,
        fields: { data: action.payload.data, isFetching: false },
      };
    case resultsActions.UPDATE_TABLE_COLUMNS: {
      return {
        ...state,
        tableColumns: action.payload.tableColumns,
      };
    }
    default:
      return state;
  }
};

export default resultsReducers;
