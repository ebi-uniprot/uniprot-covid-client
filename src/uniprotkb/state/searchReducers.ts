import { ActionType } from 'typesafe-actions';
import * as searchActions from './searchActions';
import { createEmptyClause, createPreSelectedClauses } from '../utils/clause';
// import createQueryString from '../utils/QueryStringGenerator';
import searchInitialState, { SearchState } from './searchInitialState';
import { Clause } from '../types/searchTypes';

export type SearchAction = ActionType<typeof searchActions>;

export const clause = (state: Clause, action: SearchAction) => {
  switch (action.type) {
    case searchActions.SELECT_SEARCH_TERM:
      return {
        ...state,
        searchTerm: action.payload.searchTerm,
        queryInput: action.payload.queryInput,
      };
    case searchActions.UPDATE_INPUT_VALUE:
      return {
        ...state,
        queryInput: {
          ...state.queryInput,
          stringValue: action.payload.value,
          id: action.payload.id,
        },
      };
    case searchActions.UPDATE_RANGE_VALUE:
      return {
        ...state,
        queryInput: {
          ...state.queryInput,
          [action.payload.from ? 'rangeFrom' : 'rangeTo']: action.payload.value,
        },
      };
    case searchActions.UPDATE_EVIDENCE:
      return {
        ...state,
        queryInput: {
          ...state.queryInput,
          evidenceValue: action.payload.value,
        },
      };
    case searchActions.UPDATE_LOGIC_OPERATOR:
      return {
        ...state,
        logicOperator: action.payload.value,
      };
    default:
      return state;
  }
};

export const searchTerms = (
  state: SearchState['searchTerms'],
  action: SearchAction
) => {
  switch (action.type) {
    case searchActions.REQUEST_SEARCH_TERMS:
      return {
        ...state,
        isFetching: true,
      };
    case searchActions.RECEIVE_SEARCH_TERMS:
      return {
        ...state,
        isFetching: false,
        data: action.payload.data,
        lastUpdated: action.payload.receivedAt,
      };
    default:
      return state;
  }
};

export const evidences = (
  state: SearchState['evidences'],
  action: SearchAction
) => {
  switch (action.type) {
    case searchActions.REQUEST_EVIDENCES:
      return {
        ...state,
        [action.payload.evidencesType]: {
          ...state[action.payload.evidencesType],
          isFetching: true,
        },
      };
    case searchActions.RECEIVE_EVIDENCES:
      return {
        ...state,
        [action.payload.evidencesType]: {
          isFetching: false,
          data: action.payload.data,
          lastUpdated: action.payload.receivedAt,
        },
      };
    default:
      return state;
  }
};

const searchReducers = (
  state: SearchState = searchInitialState,
  action: SearchAction
): SearchState => {
  switch (action.type) {
    case searchActions.SELECT_SEARCH_TERM:
    case searchActions.UPDATE_INPUT_VALUE:
    case searchActions.UPDATE_RANGE_VALUE:
    case searchActions.UPDATE_EVIDENCE:
    case searchActions.UPDATE_LOGIC_OPERATOR:
      return {
        ...state,
        clauses: state.clauses.map((c) => {
          if (c.id !== action.payload.clauseId) {
            return c;
          }
          return clause(c, action);
        }),
      };
    case searchActions.SUBMIT_ADVANCED_QUERY:
      return {
        ...state,
        // queryString: createQueryString(state.clauses),
      };
    case searchActions.SET_PRE_SELECTED_CLAUSES:
      return {
        ...state,
        clauses: createPreSelectedClauses(),
      };
    case searchActions.ADD_CLAUSE:
      return {
        ...state,
        clauses: [...state.clauses, createEmptyClause()],
      };
    case searchActions.REMOVE_CLAUSE:
      if (state.clauses.length === 1) {
        return {
          ...state,
          clauses: [createEmptyClause()],
        };
      }
      return {
        ...state,
        clauses: state.clauses.filter((c) => c.id !== action.payload.clauseId),
      };
    case searchActions.UPDATE_CLAUSES:
      return {
        ...state,
        clauses: action.payload.clauses,
      };
    case searchActions.REQUEST_SEARCH_TERMS:
    case searchActions.RECEIVE_SEARCH_TERMS:
      return {
        ...state,
        searchTerms: searchTerms(state.searchTerms, action),
      };
    case searchActions.REQUEST_EVIDENCES:
    case searchActions.RECEIVE_EVIDENCES:
      return {
        ...state,
        evidences: evidences(state.evidences, action),
      };
    case searchActions.UPDATE_QUERY_STRING:
      return {
        ...state,
        // queryString: action.payload.queryString,
      };
    case searchActions.RESET_SEARCH_INPUT:
      return {
        ...state,
        // queryString: '',
      };
    default:
      return state;
  }
};

export default searchReducers;
