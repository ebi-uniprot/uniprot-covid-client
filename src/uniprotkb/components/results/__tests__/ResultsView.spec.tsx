import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import ResultsView from '../ResultsView';
import { ViewMode } from '../../../state/resultsInitialState';
import { render } from '@testing-library/react';
import data from './__mocks__/results.json';

describe('ResultsView component', () => {
  test('should render table', () => {
    const props = {
      viewMode: ViewMode.TABLE,
      tableColumns: ['accession'],
      selectedEntries: [],
      results: data.results,
      sort: { column: 'accession', direction: 'descend' },
      handleEntrySelection: jest.fn(),
      handleHeaderClick: jest.fn(),
      handleLoadMoreRows: jest.fn(),
      totalNumberResults: 2,
      sortColumn: null,
      sortDirection: null,
    };
    const { asFragment } = render(
      <Router>
        <ResultsView {...props} />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render cards', () => {
    const props = {
      viewMode: ViewMode.CARD,
      tableColumns: ['accession'],
      selectedEntries: [],
      results: data.results,
      sort: { column: 'accession', direction: 'descend' },
      handleEntrySelection: jest.fn(),
      handleHeaderClick: jest.fn(),
      handleLoadMoreRows: jest.fn(),
      totalNumberResults: 2,
      sortColumn: null,
      sortDirection: null,
    };
    const { asFragment } = render(
      <Router>
        <ResultsView {...props} />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
