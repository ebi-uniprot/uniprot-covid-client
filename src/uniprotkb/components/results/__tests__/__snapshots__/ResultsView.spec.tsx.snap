// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsView component should render cards 1`] = `
<DocumentFragment>
  <div
    class="loader-container"
  >
    <svg
      class="loader"
      height="100"
      viewBox="0 0 100 100"
      width="100"
    >
      <g
        fill="none"
        fill-rule="evenodd"
      >
        <path
          d="M9 50c6.833-10.667 13.667-16 20.5-16S43.167 39.333 50 50s13.667 16 20.5 16S84.167 60.667 91 50"
          stroke="#3ca3ca"
          stroke-linecap="square"
          stroke-width="2"
        />
        <circle
          cx="9"
          cy="50"
          fill="#3ca3ca"
          r="6"
        />
        <circle
          cx="30"
          cy="34"
          fill="#3ca3ca"
          r="6"
        />
        <circle
          cx="50"
          cy="50"
          fill="#3ca3ca"
          r="6"
        />
        <circle
          cx="70"
          cy="66"
          fill="#3ca3ca"
          r="6"
        />
        <circle
          cx="91"
          cy="50"
          fill="#3ca3ca"
          r="6"
        />
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`ResultsView component should render table 1`] = `
<DocumentFragment>
  <div
    class="loader-container"
  >
    <svg
      class="loader"
      height="100"
      viewBox="0 0 100 100"
      width="100"
    >
      <g
        fill="none"
        fill-rule="evenodd"
      >
        <path
          d="M9 50c6.833-10.667 13.667-16 20.5-16S43.167 39.333 50 50s13.667 16 20.5 16S84.167 60.667 91 50"
          stroke="#3ca3ca"
          stroke-linecap="square"
          stroke-width="2"
        />
        <circle
          cx="9"
          cy="50"
          fill="#3ca3ca"
          r="6"
        />
        <circle
          cx="30"
          cy="34"
          fill="#3ca3ca"
          r="6"
        />
        <circle
          cx="50"
          cy="50"
          fill="#3ca3ca"
          r="6"
        />
        <circle
          cx="70"
          cy="66"
          fill="#3ca3ca"
          r="6"
        />
        <circle
          cx="91"
          cy="50"
          fill="#3ca3ca"
          r="6"
        />
      </g>
    </svg>
  </div>
</DocumentFragment>
`;
